scripts/lib/CIME/case/check_input_data.py

```diff
@@ -205,13 +244,57 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
                             no_files_missing = False
 
                             if (download):
+                                if firstdownload:
+                                    # Get the md5 checksum file
+                                    got_chksum = _download_checksum_file(server, input_data_root)
+                                    firstdownload = False
+                                isdirectory=rel_path.endswith(os.sep)
                                 no_files_missing = _download_if_in_repo(server, input_data_root, rel_path.strip(os.sep),
-                                                               isdirectory=rel_path.endswith(os.sep))
+                                                                        isdirectory=isdirectory)
+                                if got_chksum and no_files_missing and not isdirectory:
+                                    verify_chksum(input_data_root,"inputdata_checksum.dat", rel_path.strip(os.sep))
                         else:
+                            if chksum:
+                                verify_chksum(input_data_root,"inputdata_checksum.dat", rel_path.strip(os.sep))
+                                logger.info("Chksum passed for file {}".format(os.path.join(input_data_root,rel_path)))
                             logging.debug("  Already had input file: '{}'".format(full_path))
 
+
+
                 else:
                     model = os.path.basename(data_list_file).split('.')[0]
                     logging.warning("Model {} no file specified for {}".format(model, description))
 
     return no_files_missing
+
+
+def verify_chksum(input_data_root, checksumfile, filename):
+    if not chksum_hash:
+        hashfile = os.path.join(input_data_root,checksumfile)
+        if not os.path.isfile(hashfile):
+            expect(False, "Failed to find or download file "+hashfile+
+                   " download from ftp:ftp.cgd.ucar.edu/cesm/inputdata_chksum.dat")
+
+        with open(hashfile) as fd:
+            lines = fd.readlines()
+            for line in lines:
+                lsplit = line.split()
+                if len(lsplit) < 8:
+                    continue
+                fname = (lsplit[7])[10:]
+                chksum_hash[fname] = lsplit[0]
+
+    chksum = md5(os.path.join(input_data_root, filename))
+    if chksum_hash:
+        expect(filename in chksum_hash.keys() and chksum == chksum_hash[filename],
```

Comment:
It seems like there are a few problems with expecting the current file to be in chksum_hash:

1. Tied in to my above comments: let's say we just downloaded a file from server B, but chksum_hash was set based on server A; then this new file would be missing from chksum_hash.

2. When running `check_input_data` with the new `--chksum` argument: My impression is that you'll end up trying to do this checksum for each server, but the file only exists on one server, so wouldn't the expect fail for all of the other servers? This case also has the problem of (1). In addition, due to the problem of using the same file name for all servers, this case has the problem that the checksum file on disk may never be for the server this file resides on (if no files need to be downloaded from that server), because it looks like existing files don't force a download of the checksum file from the server.

3. Can we assume that the checksum file on the server is updated near-instantaneously whenever a file is added to the server? If not, we could have a situation where you upload a file from one machine, then a few minutes later go to download it on a different machine, and get an error because the new file is not yet in the server's checksum file.

Reply:
In addition, if you do keep the check of filename in chksum_hash.keys(), you should make this a separate check, rather than checking both conditions at once: These two issues (missing file name in the chksum file and mismatching chksum) are fundamentally different problems, and it's confusing to have them reported with the same error.